### ISSUE TITLE
Feature Request: Access to last change date of an item (for iobroker driver)

### ISSUE THREAD

**Issue by tukey42** [Martin-Gleiss/smartvisu#983]
It would be great, if the last change date of an item can be accessed (especially for the iobroker driver).

**Comment by uwe5** [comment]
I would like that too!

The timestamp arrives in the ioBroker adapter (io_iobroker.js) here:
stateChanged: function(item, state) {
...
// state.lc: TIMESTAMP of last change
...
}

Here you would have to somehow send state.lc instead of state.val to the widget, perhaps via an attribute as switch to use state.lc instead of state.val.
I wanted to include this concept here, but I don't know enough about it.

State is the actual value of the data point and presented by javascript object:
{
    val: VALUE, 
    ack: ACKNOWLEDGED, 
    ts: TIMESTAMP, // could be converted into time with "new Date(state.ts)" (In older version of js-controller - "new Date(state.ts * 1000)")
    lc: TIMESTAMP of last change, 
    from: ADAPTER_NAME, 
    q: QUALITY
}

greetings Uwe

**Comment by wvhn** [comment]
I am already working on it. 

Since we have the keyword „property“ for item properties already in smarthomeNG and the templatechecker can check this I would use this here, as well. 
So to visualize the items last changed property you would enter `myItem.property.lc` as parameter in the widget which then is filled with myItem.state.lc in the driver. 
To simplify the procedure, I would map all item properties if any one is requested on the actual page for that item. Otherwise, only myItem.state.val is mapped like in the actual driver.

Does that fulfill the requirements?

regards
Wolfram

**Comment by tukey42** [comment]
Yes, that would be great!

Best Regards, Stefan

**Comment by wvhn** [comment]
OK. Can you test the attached driver for me, after renaming it from .txt to .js? I ran some dry tests but have no ioBroker system for validation.
[io_iobroker_new.txt](https://github.com/Martin-Gleiss/smartvisu/files/15370318/io_iobroker_new.txt)
Request the item last changed property as `myItem.property.lc`. You can display the time with `basic.print('','myItem.property.lc','time','VAR*1000')` if the timestamp is in seconds.

Regards
Wolfram

**Comment by tukey42** [comment]
I did a quick test. Unfortunately I got a lot of the following error message in the browser console:

TypeError: io.openItems.removeEntry is not a function. (In 'io.openItems.removeEntry(item)', 'io.openItems.removeEntry' is undefined)

Regards
   Stefan

**Comment by wvhn** [comment]
Ah. Sorry! I forgot that I had implemented already a new  function in the driver which is not compatible with v3.4. Can you run the test in the develop version of smartVISU? 
Alternatively you could try to put the following code in visu.js in the folder of your pages:
```
/**
 * workaround for v3.4.  Delete after upgrading to a later version
 * 
 * Removes the first occurrence of a given array element
 */
 if(!Array.prototype.removeEntry) {
	 Array.prototype.removeEntry = function(value) {
		if (!value)
			return false;
		var index = this.indexOf(value);
		if (index > -1)
			this.splice(index,1);
		return true;
	 }
	// Hide method from for-in loops
	Object.defineProperty(Array.prototype, "removeEntry", {enumerable: false});
}
```

Regards 
Wolfram

**Comment by tukey42** [comment]
Many thanks, my first test works!
If I just print the timestamp as integer `basic.print('','myItem.property.lc')` the correct Unix timestamp is shown.
If I use `basic.print('','myItem.property.lc','time','VAR*1000')` a totally wrong time is shown, but this seems to be an issue with basic.print itself.

Best Regards
    Stefan

**Comment by wvhn** [comment]
Since @uwe5 wrote that the controller versions may differ in the timestamps unit (ms or s) I added the formula `VAR * 1000` in case the timestamp is in seconds.  I guess your version is already in ms. Just omit the formula and then the time should be correct. 

Regards
Wolfram

**Comment by tukey42** [comment]
It does not show the right time  with or without *1000.
The timestamp is in ms:
1716196335350 (Mon May 20 2024 11:12:15 GMT+0200 (Mitteleuropäische Sommerzeit))
With *1000 it prints: 13:15 (time) or 05.02.56354 13:15 (long)
Without *1000 it prints: aN:aN

I will test the new version of basic.print later. I am on 3.3.1.

Best Regards
    Stefan

**Comment by wvhn** [comment]
I had been editing my comment above  after testing with v3.4. Please review. 

Regards 
Wolfram

**Comment by wvhn** [comment]
I tested it in v3.3.1. `VAR*1` works there as a workaround. The multiplication with 1 converts the timestamp string into a number which is required for conversion into a time/date.

**Comment by wvhn** [comment]
implemented with #984
